src/async_events.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2020 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0 or MIT
+//
+
+use futures_core::Stream;
+use std::os::unix::prelude::AsRawFd;
+use std::path::Path;
+use tokio::fs;
+use tokio::io::AsyncReadExt;
+
+use crate::error::ErrorKind::*;
+use crate::error::*;
+
+// notify_on_oom returns channel on which you can expect event about OOM,
+// if process died without OOM this channel will be closed.
+pub async fn notify_on_oom_v2(key: &str, dir: &Path) -> Result<impl Stream<Item = String>> {
+    register_memory_event(key, dir, "memory.oom_control", "").await
```

Comment:
The oom event mechanism is quite different in cgroups v1 and cgroups v2, e.g., There is no file memory.oom_control in cgroups v2

FYI: https://github.com/kata-containers/kata-containers/blob/main/src/agent/rustjail/src/cgroups/notifier.rs#L61

Reply:
Ok. Good to know ;-)

I just took the sync impl from `cgroups-rs` as reference which doesn't distinguish between v1 and v2. I'll check your link.

Reply:
@flxo Have you test on cgroups v2?

Reply:
@Tim-Zhang No - just on v1.

Reply:
@Tim-Zhang @jodh-intel 

Turns out that v2 is a bigger story:

1.[events::notify_on_oom_v2]( https://docs.rs/cgroups-rs/0.2.6/cgroups_rs/events/fn.notify_on_oom_v2.html) is broken (unimplemented) for v2.
2. `async_events::notify_on_oom_v2` of this PR is just a async copy of `events::notify_on_oom_v2` -> broken
3. `[memory::MemController::disable_oom_killer](https://docs.rs/cgroups-rs/0.2.6/cgroups_rs/memory/struct.MemController.html#method.disable_oom_killer) is a v1 impl only and not guarded for v2. This fn should at least return an error.

My proposal would be:

1. Remove `events::notify_on_oom_v2` - it probably not used by anybody either
2. Add the `v1` async version of this PR
3. In a follow up PR add sync and async v2 support

I could update this PR accordingly. Thanks!

Reply:
Hello. Anyone have had time to take a look at this? I'm still willing to work on this with appropriate support.